### ISSUE TITLE
ItemRepSelector should prefer unprocessed items

### ISSUE THREAD

**Issue by ddfreyne** [nanoc/nanoc#977]
The `ItemRepSelector` does not pick items to recompile in the most efficient order.

Theory: If an item is known to have a dependency (because attempting to compiling it raised an `UnmetDependency`), it is more likely to have dependencies than an item that has not yet been processed yet.

For example, a blog archive page, a tag page, … will collect the compiled content of many items, and thus has many dependencies. It makes sense for Nanoc to attempt compiling all other items before re-attempting compiling the item that generated the dependency.

CC @gpakosz

**Comment by ddfreyne** [comment]
Another useful enhancement is to repurpose the information in the dependency graph of the previous run. That’s something I’ve been meaning to do, but never got around to doing…

**Comment by ddfreyne** [comment]
The cause of this problem is illustrated by the new [`ItemRepSelector` spec](https://github.com/nanoc/nanoc/blob/master/spec/nanoc/base/services/item_rep_selector_spec.rb): in the last example, the item rep selector yields the same item `:a` over and over again:

```ruby
expect(tentatively_yielded).to eq [:a, :b, :a, :c, :a, :d, :a, :e, :a]
```

The reason for this is that as of Nanoc 4.3.7, the initial set of item reps to recompile that is passed to the item rep selector is not *every* item, but only the outdated ones. Because of this, the item rep selector does not have enough information to realise that there are other items reps that might need to be recompiled before attempting to recompile `:a`, and so it has to recompile `:a` over and over again to find out more and more dependencies.

Repurposing the dependency graph from the previous run, as mentioned before, would avoid this problem.

**Comment by ddfreyne** [comment]
Fix in #985.
